### ISSUE TITLE
Np level detection

### ISSUE THREAD

**Issue by Dante0909** [MathewSachin/Fate-Grand-Automata#334]
Hello,
I thought of this feature because I'm going to farm Gilfest and I need a friend np5 jarcher.
I reckon adding a np level check should be doable? It seems like a very beneficial addition.
Thank you
Dante

**Comment by MathewSachin** [comment]
Seems like this isn't as simple as I thought initially now that I looked at many support servants.

These are the challenges:
1. Length of NP name causes the NP level number to be at different positions.
2. The NP level can be in 3 different colors depending on Not-friend/friend/Interlude or Rankup done.
3. I don't recall if NPs have numbers in their name which could be wrongly detected as NP levels.
4. The NP name shifts depending on whether master message is set. This one is simple to handle, just added here so I don't forget.

<img src="https://user-images.githubusercontent.com/10654537/91636628-bdf42000-ea1f-11ea-87f3-03bccdc9ec9e.jpg">
<img  src="https://user-images.githubusercontent.com/10654537/91636630-bfbde380-ea1f-11ea-9b54-c05a65e91a2a.jpg">
<img src="https://user-images.githubusercontent.com/10654537/91636632-c0567a00-ea1f-11ea-8391-d1d4a865f14c.jpg">
<img src="https://user-images.githubusercontent.com/10654537/91636633-c0ef1080-ea1f-11ea-97b9-8bae410fec20.jpg">

**Comment by tchofy** [comment]
One servant I can think of that has numbers on their NP is summer Kiyo (Dojoji Bell, Method 108 - Fire Dragon Mower), seems like the detection would have to include the "Lv. "

**Comment by reconman** [comment]
Another challenge is differing font size:
![Screenshot_20200829-171119](https://user-images.githubusercontent.com/12360257/91639984-d6ae0700-ea1a-11ea-8975-6deeb455d0c2.jpg)

Just compare both Lv. 5

**Comment by MathewSachin** [comment]
@tchofy Ah, I'm dumb, nice catch, Lv is present for everyone.

So, only checking `Lv. 5` in both white and yellow should be good enough (at-least for NA).
When NP 5 is asked for, we should assume Friends only.

@reconman In that case, this won't work without text recognition 😢.

**Comment by Dante0909** [comment]
Hmm these seem like tough challenges.
Perhaps as an alternative, you could only check for np5(and friends obviously, otherwise you can't use np)?
So you just check if there's yellow for like the first letter. I don't know if it's possible, but just suggesting the idea.

**Comment by reconman** [comment]
@Dante0909 Image recognition is always done in black and white so far.

**Comment by MathewSachin** [comment]
@Dante0909 Missed your last comment. The colour isn't the problem here, it's the varying font size like @reconman pointed out which we don't have a way around right now.

**Comment by TryBane** [comment]
This type of problem is the reason I initially implemented the Friend Name feature. Since it is annoying and difficult to deal with all of these problems, I went with a simple route where you can just use friend names if they are the ones with the NP levels you need. If it is possible for you to be in contact with them personally, you can ask them to keep that servant up. Alternatively, the contributors can always mix the Friend Name feature with the Servant/CE preferred feature and allow a cross examination between the Friend Names you selected and if they have the Servant/CE you are looking for that you know they have.

This shouldn't be too difficult to implement.

**Comment by LouizFC** [comment]
Would OCR be an overkill here?

**Comment by MathewSachin** [comment]
@LouizFC with OCR, the location of "Lv. X" would need to be specified, which varies depending on how long the NP's name is.
If the rest of the text is not cropped out correctly, OCR would give bad results, more so for Chinese, Japanese, Korean texts if we're using an English OCR.
I had tried using OCR for Skill level checks but the results weren't good, so I resorted to Image matching there too. But, image matching won't work here since the font size varies depending on the length of NP name.

**Comment by LouizFC** [comment]
@MathewSachin I forgot about Chinese, Japanese and Korean texts, sorry. I have been playing with some lightweight neural network image recognition at work, when I get some time I will run some tests. Do you know how I can get my hands on some text samples from the mentioned languages? (aside from creating a new account on each region, of course. But if that is the only option I will go with that too)

**Comment by LouizFC** [comment]
I experimented with some custom Tesseract training for a different subject today.
When I get some time I pretend to:

1. Collect some FGO samples from videos and test.
2. Test on a simple PC implementation
3. Port the implementation to an android friendly module.

**Comment by reconman** [comment]
> When I get some time I pretend to:

Do you mean "plan to"?

**Comment by LouizFC** [comment]
@reconman Exactly. Sorry about that. In my native language there verb called "pretendo", which translates to "intend to" in english, I though it was a cognate, but looking up now I was wrong. Thanks for the correction

**Comment by the3dsandwich** [comment]
Just saw you mentioned Chinese. I have a TW account, can I help in any way?

**Comment by LouizFC** [comment]
> Just saw you mentioned Chinese. I have a TW account, can I help in any way?

@the3dsandwich thank you for wanting to help. I need some screenshots from your support list of some specific nps:
- Kyohime summer - because it has numbers in the name
- Anne & Mary Summer - same reason as Kyohime
- Assassin Shiki - very long name that distorts the np "Lv" text.

**Comment by the3dsandwich** [comment]
Are you marching the English part of the names? Chinese numbers are drastically different from normal numbers that I don't think it'll affect image matching. Also I'll be on the lookout for long NP names (assassin Shiki's NP name is on the short side)

**Comment by the3dsandwich** [comment]
![image](https://user-images.githubusercontent.com/32169552/107883894-b58b9500-6f2c-11eb-8c63-a2ff145c701b.jpeg)
![image](https://user-images.githubusercontent.com/32169552/107883902-c3d9b100-6f2c-11eb-98d7-aabb7cbea9ea.jpeg)
I know you asked for the lancer version but none of my friends have a lancer kiyo atm and regular kiyo also has a number in her NP. 
![image](https://user-images.githubusercontent.com/32169552/107883906-c76d3800-6f2c-11eb-827e-1ea6d5ab40c0.jpeg)
![image](https://user-images.githubusercontent.com/32169552/107883908-c9cf9200-6f2c-11eb-98dc-a06f7e3dfe4d.jpeg)
![image](https://user-images.githubusercontent.com/32169552/107883910-cc31ec00-6f2c-11eb-9e95-8b1a814df442.jpeg)
Stella in TW server also have a number 1 in it. 

As you may see now TW NP names are generally short, but I'll be on the lookout for long NP names. Also I think Chinese numbers is different enough to not cause an in,age recognition problem?

**Comment by MathewSachin** [comment]
If TW NP names are always short, simple image recognition should be enough.
If only other servers could be the same case 😢 .

> Also I think Chinese numbers is different enough to not cause an image recognition problem?

Image matching would indeed fail by just minor font changes, but, since LouizFC is trying OCR, I think number font changes might not be a problem.

**Comment by LouizFC** [comment]
@the3dsandwich thank you very much for the samples. I will try to manually "resize" the "Lv" text  just in case.

@MathewSachin before I work on this, is there any objection in using Tesseract in the app? because if yes I can try other lightweight approaches first

**Comment by MathewSachin** [comment]
I'm fine with Tesseract if it gets the job done. I believe the APK size increase won't be as drastic as when adding x64 libraries (#624).

I had tried Google on device ML APIs for detecting skill levels before I resorted to image matching because the results weren't satisfactory. Well, I'm just not good at ML stuff in general.

**Comment by LouizFC** [comment]
@MathewSachin I am also not good at ML stuff, I began studying it in the past month. I got interest in this issue as a learning oportunity

**Comment by danilaml** [comment]
It feels like there should be good solution for OCR-ing numbers (from a known font) on-device, but even ignoring that, can't we at least just get NP5 detection? This would already cover a lot of cases (when farming relies on maxed support DPS or 4*) and you only need to detect that font is yellow.

**Comment by danilaml** [comment]
Btw, did a quick and dirty test looking for `Lv. ?(\d)` and tesseract detectText was reasonably okay although it sometimes would confuse `1` and `T`, but googles mlkit appeared to be way more accurate. Perhaps it's time reevaluate the feasibility.
